pkg/discovery/dtofactory/application_entity_dto_builder.go

```diff
@@ -43,130 +42,184 @@ func NewApplicationEntityDTOBuilder(sink *metrics.EntityMetricSink) *application
+				glog.Errorf("Failed to build Application entityDTO based on application %s: %s", displayName, err)
+				continue
+			}
+			result = append(result, entityDTO)
 		}
-		entityDTOBuilder.BuysCommodities(commoditiesBought)
+	}
 
-		// entities' properties.
-		properties := builder.getApplicationProperties(pod)
-		entityDTOBuilder.WithProperties(properties)
+	return result, nil
+}
 
-		if !util.Monitored(pod) {
-			entityDTOBuilder.Monitored(false)
-		}
+func (builder *applicationEntityDTOBuilder) getTransactionUsedValue(pod *api.Pod) float64 {
+	key := util.PodKeyFunc(pod)
+	etype := task.PodType
+	rtype := metrics.Transaction
+	mtype := metrics.Used
+	metricsId := metrics.GenerateEntityResourceMetricUID(etype, key, rtype, mtype)
 
-		appType := util.GetAppType(pod)
-		entityDTOBuilder.ApplicationData(&proto.EntityDTO_ApplicationData{
-			Type: &appType,
-		})
+	usedMetric, err := builder.metricsSink.GetMetric(metricsId)
+	if err != nil {
+		glog.V(3).Infof("failed to get Pod[%s] transaction usage: %v", key, err)
+		return 0.0
+	}
 
-		// build entityDTO
-		entityDTO, err := entityDTOBuilder.Create()
-		if err != nil {
-			glog.Errorf("Failed to build Application entityDTO based on application %s: %s", displayName, err)
-			continue
-		}
+	return usedMetric.GetValue().(float64)
+}
+
+func (builder *applicationEntityDTOBuilder) getCPUUsedValue(etype task.DiscoveredEntityType, key string) float64 {
+	rtype := metrics.CPU
+	mtype := metrics.Used
+	metricsId := metrics.GenerateEntityResourceMetricUID(etype, key, rtype, mtype)
 
-		result = append(result, entityDTO)
+	usedMetric, err := builder.metricsSink.GetMetric(metricsId)
+	if err != nil {
+		glog.Warningf("failed to get CPU usage for %v: %v", etype, err)
+		return 0.0
 	}
 
-	return result, nil
+	return usedMetric.GetValue().(float64)
 }
 
-// Build the sold commodityDTOs by each application.
-// An application sells: Transaction commodity.
-func (builder *applicationEntityDTOBuilder) getApplicationCommoditiesSold(pod *api.Pod) ([]*proto.CommodityDTO, error) {
+// calculate app Transaction Use = (pod.Transaction.used * container.CPU.used)/pod.CPU.used
+func (builder *applicationEntityDTOBuilder) getAppTransactionUsage(index int, pod *api.Pod) float64 {
 
-	// Cluster commodity.
-	clusterMetricUID := metrics.GenerateEntityStateMetricUID(task.ClusterType, "", metrics.Cluster)
-	clusterInfo, err := builder.metricsSink.GetMetric(clusterMetricUID)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get %s used for current Kubernetes Cluster", metrics.Cluster)
+	podTransactionUsage := builder.getTransactionUsedValue(pod)
+	if podTransactionUsage < 2.0 || len(pod.Spec.Containers) < 2 {
+		return podTransactionUsage
 	}
-	clusterCommodityKey, ok := clusterInfo.GetValue().(string)
-	if !ok {
-		return nil, errors.New("Failed to get cluster ID")
+
+	//1. get pod CPU usage
+	podkey := util.PodKeyFunc(pod)
+	podUsage := builder.getCPUUsedValue(task.PodType, podkey)
+	if podUsage < 0.001 {
```

Comment:
Why we need this?

Reply:
Maybe you can distinguish different container by different port number. But I don't know how much effort would it be.

Reply:
I think distributing transactions based on CPU is better.

The reason is that:
    (1) Some sidecar containers may not expose Port. For example, a sidecar container only push logs onto a log server;
     (2) Some sidecar containers may be very important.  For example, there are two containers, one is for network interaction; the other is responsible for computation, and sends the results back to the first one. In this case, it seems no outbound network activity for the second container.  But we'd better to make sure it has enough resources.

In these assumptions, I seems that

Reply:
After a discussion with Enlin, we can simply equally distributed the Pod transactions to the hosted containers.  It does not matter which container get more transactions, since transaction is used to drive Pod to be cloned, not the containers.